lib/puppet/datatypes.rb

```diff
@@ -0,0 +1,190 @@
+#       sender.send_text_from(@name, text)
+#     end
+#   end
+#
+# Then the type declaration can look like this:
+#
+#   Puppet::DataTypes.create_type('Auth::User') do
+#     interface <<-PUPPET
+#       attributes => {
+#         name => String,
+#         email => String,
+#         year_of_birth => Integer,
+#         age => { type => Integer, kind => derived }
+#       },
+#       functions => {
+#         send_text => Callable[Sender, String[1]]
+#       }
+#       PUPPET
+#
+#     # This require is optional and only needed in case
+#     # the implementation is not loaded by other means.
+#     require 'auth/utils/user.rb'
+#
+#     implementation_class Auth::Utils::User
+#   end
+#
+module Puppet::DataTypes
+  def self.create_type(type_name, &block)
+    # Ruby < 2.1.0 does not have method on Binding, can only do eval
+    # and it will fail unless protected with an if defined? if the local
+    # variable does not exist in the block's binder.
+    #
+    begin
+      loader = block.binding.eval('loader_injected_arg if defined?(loader_injected_arg)')
+      create_loaded_type(type_name, loader, &block)
+    rescue StandardError => e
+      raise ArgumentError, _("Data Type Load Error for type '%{type_name}': %{message}") % {type_name: type_name, message: e.message}
+    end
+
+  end
+
+  def self.create_loaded_type(type_name, loader, &block)
+    builder = TypeBuilder.new(type_name.to_s)
+    api = TypeBuilderAPI.new(builder).freeze
+    api.instance_eval(&block)
+    builder.create_type(loader)
+  end
+
+  class TypeBuilder
+    attr_accessor :interface, :implementation, :implementation_class
+
+    def initialize(type_name)
+      @type_name = type_name
+      @implementation = nil
+      @implementation_class = nil
+    end
+
+    def create_type(loader)
+      raise ArgumentError, _('a type must have an interface') unless @interface.is_a?(String)
+      created_type = Puppet::Pops::Types::PObjectType.new(
+        @type_name,
+        Puppet::Pops::Parser::EvaluatingParser.new.parse_string("{ #{@interface} }").body)
+
+      if !@implementation_class.nil?
+        if @implementation_class < Puppet::Pops::Types::PuppetObject
+          @implementation_class.instance_eval do
+            include Puppet::Pops::Types::PuppetObject
+            @_pcore_type = created_type
+
+            def self._pcore_type
+              @_pcore_type
+            end
+          end
+        else
+          Puppet::Pops::Loaders.implementation_registry.register_implementation(created_type, @implementation_class, loader)
+        end
+        created_type.implementation_class = @implementation_class
+      elsif !@implementation.nil?
+        created_type.implementation_override = @implementation
+      end
+      created_type
+    end
+
+    def has_implementation?
+      !(@implementation_class.nil? && @implementation.nil?)
+    end
+  end
+
+  # The TypeBuilderAPI class exposes only those methods that the builder API provides
+  # @api private
```

Comment:
why is this class `@api private` when it is a public API? Isn't it just the `initialize` that is private?